flight_sql/accessors/timestamp_array_accessor.cc

```diff
@@ -72,11 +83,18 @@ TimestampArrayFlightSqlAccessor<TARGET_TYPE, UNIT>::MoveSingleCell_impl(
     ColumnBinding *binding, int64_t arrow_row, int64_t cell_counter,
     int64_t &value_offset, bool update_value_offset,
     odbcabstraction::Diagnostics &diagnostics) {
+  // Times less than the minimum integer number of seconds that can be represented
+  // for each time unit will not convert correctly.  This is mostly interesting for
+  // nanoseconds as timestamps in other units are outside of the accepted range of
+  // Gregorian dates.
   auto *buffer = static_cast<TIMESTAMP_STRUCT *>(binding->buffer);
 
   int64_t value = this->GetArray()->Value(arrow_row);
   const auto divisor = GetConversionToSecondsDivisor(UNIT);
-  const auto converted_result_seconds = value / divisor;
+  const auto converted_result_seconds =
+    (value < 0)
+    ? ((value - (divisor - 1)) / divisor)
```

Comment:
This formula needs an explanation.

Reply:
Does that parse ok?

Reply:
Can we make GetConversionToSecondsDivisor() take and return constexprs and make divisor a constexpr? (This should help to both optimize out the switch and treat the divisor as a constant)?

Reply:
I am inclined to factor some of that as the more obvious floor_div() and the portable modulo which is how it would be written if C++ was useful :)  Also any reason for constexpr functions instead of just inline?

Reply:
inline is just a suggestion. constexpr has a stronger guarantee (which is that constants-only will be used), making the function more friendly for constant folding.